src/native/clr/host/gc-bridge.cc

```diff
@@ -1,44 +1,315 @@
+	env->DeleteGlobalRef (handle);
+}
+
+void GCBridge::cleanup_after_java_collection (MarkCrossReferencesArgs* cross_refs) noexcept
+{
+#if DEBUG
+	int total = 0;
+	int alive = 0;
+#endif
 
-	// Call back into managed code
-	bridge_processing_finish_callback (crossRefs);
+	// try to switch back to global refs to analyze what stayed alive
+	for (size_t i = 0; i < cross_refs->ComponentCount; i++) {
+		StronglyConnectedComponent *scc = &cross_refs->Components [i];
+		for (size_t j = 0; j < scc->Count; j++) {
+#if DEBUG
+				total++;
+#endif
+				take_global_ref (scc->Contexts [j]);
+			}
+	}
+
+	// clear the cross references on any remaining items
+	for (size_t i = 0; i < cross_refs->ComponentCount; i++) {
+		StronglyConnectedComponent *scc = &cross_refs->Components [i];
+		bool is_alive = false;
+
+		for (size_t j = 0; j < scc->Count; j++) {
+			JniObjectReferenceControlBlock *control_block = scc->Contexts [j]->control_block;
+
+			if (control_block != nullptr) {
+#if DEBUG
+				alive++;
+#endif
+				if (j > 0) {
+					abort_unless (is_alive, [&i] { return detail::_format_message ("Bridge SCC at index %d must be alive", i); });
+				}
+
+				is_alive = true;
+				if (control_block->refs_added > 0) {
+					clear_references (control_block->handle);
+					control_block->refs_added = 0;
+				}
+			} else {
+				abort_unless (!is_alive, [&i] { return detail::_format_message ("Bridge SCC at index %d must NOT be alive", i); });
+			}
+		}
+	}
+
+#if DEBUG
+	log_info (LOG_GC, "GC cleanup summary: {} objects tested - resurrecting {}.", total, alive);
+#endif
+}
+
+void GCBridge::bridge_processing () noexcept
+{
+	abort_unless (bridge_processing_started_callback != nullptr, "GC bridge processing started callback is not set");
+	abort_unless (bridge_processing_finished_callback != nullptr, "GC bridge processing finished callback is not set");
+
+	env = OSBridge::ensure_jnienv ();
+
+	while (true) {
+		bridge_processing_semaphore.acquire ();
+		std::unique_lock<std::shared_mutex> lock (processing_mutex);
+
+		MarkCrossReferencesArgs* args = &GCBridge::shared_cross_refs;
+		log_mark_cross_references_args_if_enabled (args);
+
+		bridge_processing_started_callback ();
+		prepare_for_java_collection (args);
+		trigger_java_gc ();
+		cleanup_after_java_collection (args);
+		bridge_processing_finished_callback (args);
+	}
+}
+
+void GCBridge::mark_cross_references (MarkCrossReferencesArgs* cross_refs) noexcept
+{
+	std::unique_lock<std::shared_mutex> lock (processing_mutex);
+
+	GCBridge::shared_cross_refs.ComponentCount = cross_refs->ComponentCount;
+	GCBridge::shared_cross_refs.Components = cross_refs->Components;
+	GCBridge::shared_cross_refs.CrossReferenceCount = cross_refs->CrossReferenceCount;
+	GCBridge::shared_cross_refs.CrossReferences = cross_refs->CrossReferences;
+
+	bridge_processing_semaphore.release ();
+}
+
+void GCBridge::wait_for_bridge_processing () noexcept
+{
+	std::shared_lock<std::shared_mutex> lock (processing_mutex);
```

Comment:
This doesn't seem correct. In theory we could obtain this mutex, before the bridge worker thread actually acquires it, doing the BP2 stage. We would need at least an additional variable to mark whether we have a bridge in progress. Probably makes sense to use a condition variable for this.